packages/testing/config/jest/api/jest.setup.js

```diff
@@ -1,191 +1,57 @@
-    fs.writeFileSync(TEARDOWN_CACHE_PATH, JSON.stringify(teardownOrder))
-  }
-}
-
-const buildScenario =
-  (it) =>
-  (...args) => {
-    let scenarioName, testName, testFunc
-
-    if (args.length === 3) {
-      ;[scenarioName, testName, testFunc] = args
-    } else if (args.length === 2) {
-      scenarioName = DEFAULT_SCENARIO
-      ;[testName, testFunc] = args
-    } else {
-      throw new Error('scenario() requires 2 or 3 arguments')
-    }
-
-    return it(testName, async () => {
-      const path = require('path')
-      const testFileDir = path.parse(global.testPath)
-      // e.g. ['comments', 'test'] or ['signup', 'state', 'machine', 'test']
-      const testFileNameParts = testFileDir.name.split('.')
-      const testFilePath = `${testFileDir.dir}/${testFileNameParts
-        .slice(0, testFileNameParts.length - 1)
-        .join('.')}.scenarios`
-      let allScenarios, scenario, result
-
-      try {
-        allScenarios = require(testFilePath)
-      } catch (e) {
-        // ignore error if scenario file not found, otherwise re-throw
-        if (e.code !== 'MODULE_NOT_FOUND') {
-          throw e
-        }
-      }
-
-      if (allScenarios) {
-        if (allScenarios[scenarioName]) {
-          scenario = allScenarios[scenarioName]
-        } else {
-          throw (
-            ('UndefinedScenario',
-            `There is no scenario named "${scenarioName}" in ${testFilePath}.js`)
-          )
-        }
-      }
-
-      const scenarioData = await seedScenario(scenario)
-      result = await testFunc(scenarioData)
-
-      return result
-    })
-  }
-
-global.scenario = buildScenario(global.it)
-global.scenario.only = buildScenario(global.it.only)
-
-global.defineScenario = defineScenario
+global.scenario = buildScenario(global.it, global.testPath)
+global.scenario.only = buildScenario(global.it.only, global.testPath)
 
 global.mockCurrentUser = (currentUser) => {
   setContext({ currentUser })
 }
 
+/**
+ *
+ * All these hooks run in the VM/Context that the test runs in since we're using "setupAfterEnv".
+ * There's a new context for each test-suite i.e. each test file
+ *
+ * The value __RWJS_TEST_PRISMA_WAS_IMPORTED is set by babel-plugin-redwood-test-db -
+ * which is only used in api jest tests - see packages/testing/config/jest/api/apiBabelConfig.js
+ *
+ * Doing this means if the db isn't used in the current test context,
+ * no need to do any of the teardown logic - allowing simple tests to run faster
+ * At the same time, if the db is used, disconnecting it in this context prevents connection limit errors.
+ * Just disconnecting db in jest-preset is not enough, because
+ * the Prisma client is created in a different context.
+ */
+
 beforeAll(async () => {
   // Disable perRequestContext for tests
   process.env.DISABLE_CONTEXT_ISOLATION = '1'
-  await configureTeardown()
+  if (globalThis.__RWJS_TEST_PRISMA_WAS_IMPORTED) {
+    await configureTeardown()
+  }
```

Comment:
I see the strategy here and this definitely works but I have a suggestion that I think would let us avoid using babel. We could just check the require cache. It resets between tests cause we're in `jest.setup.js`, and we could just know if db was imported or not pretty simply:

```suggestion
  const shouldTeardown = Object.keys(require.cache).some((mod) => {
    return (
      mod === `${apiSrcPath}/lib/db.js` || mod === `${apiSrcPath}/lib/db.ts`
    )
  })

  if (shouldTeardown) {
    await configureTeardown()
  }
```

Reply:
Good eye - and this would totally be the most logical thing to do! Unfortunately, if you console log `require.cache` here it's always undefined (or maybe empty... don't remember). 

I don't fully understand what exactly is happening under the hood - but remember one of the "rules" we learnt was that inside the tests require cache isn't used, which is why we used to get unbounded memory growth! 

Another interesting thing to note here though... in `src/lib/db` we instatiate db with `new Prisma Client`. So under normal node execution, thanks to require cache it doesn't keep creating new prisma clients (and therefore new connections to the db). In theory - inside the test VM, because require.cache isn't used - you would think requiring src/lib/db (as we do in teardown) would create a new instance of prisma right? Nope 😮 - it reuses the same instance. Would love your thoughts on why you think this is happening.

The way I validated this was to check the number of connections to my postgres test db. If you comment out the teardown code (where it does the disconnect) you see that it keeps increasing till it hits the limit. With the disconnect in there, it maintains 2 connections (1 for jest-preset, 1 for the test VM) - until the tests finish.

Reply:
I think the nuance is when you check the require cache. You have to check it in the lifecycle hooks (`beforeAll`, etc)—only then will it have imported the code for the tests it's going to run. Here's a tape of me explaining what I think is happening by stepping through `yarn rw test api`: https://s.tape.sh/tjCetPbZ

> Another interesting thing to note here though... in src/lib/db we instatiate db with new Prisma Client. So under normal node execution, thanks to require cache it doesn't keep creating new prisma clients (and therefore new connections to the db). In theory - inside the test VM, because require.cache isn't used - you would think requiring src/lib/db (as we do in teardown) would create a new instance of prisma right? Nope 😮 - it reuses the same instance. Would love your thoughts on why you think this is happening.

I'll think about this next!

Reply:
Thanks for double checking here, I've now removed the babel stuff and am using the require cache like you mentioned. I'm sure I tried this as the first pass though 🤷